File: ci/kotlin/ExecutorBindingsStub.kt
package ptpsim.ci

import uniffi.camera_protocol_ffi.BleExecutorTransport
import uniffi.camera_protocol_ffi.CccdMode
import uniffi.camera_protocol_ffi.ConnectionActivityEvent
import uniffi.camera_protocol_ffi.ConnectionActivityFailure
import uniffi.camera_protocol_ffi.ConnectionActivityObserver
import uniffi.camera_protocol_ffi.ConnectionActivityRetry
import uniffi.camera_protocol_ffi.ExecutorException
import uniffi.camera_protocol_ffi.ExecutorStepFailureKind
import uniffi.camera_protocol_ffi.Predicate
import uniffi.camera_protocol_ffi.PredicateOp
import uniffi.camera_protocol_ffi.PtpExecutorTransport
import uniffi.camera_protocol_ffi.PtpSessionOpenResult
import uniffi.camera_protocol_ffi.SocketRole
import uniffi.camera_protocol_ffi.Step
import uniffi.camera_protocol_ffi.StepObserver
import uniffi.camera_protocol_ffi.StepReport

class ExecutorBindingsStub : BleExecutorTransport {
    override suspend fun connect() = Unit

    override suspend fun awaitDisconnect() = Unit

    override suspend fun requestMtu(mtu: UShort): UShort = mtu

    override suspend fun ensureServicesDiscovered() = Unit

    override suspend fun read(characteristic: String): ByteArray = byteArrayOf()

    override suspend fun write(characteristic: String, value: ByteArray) = Unit

    override suspend fun writeWithNotificationFence(
        characteristic: String,
        value: ByteArray,
        notificationCharacteristic: String,
    ) = Unit

    override suspend fun subscribe(characteristic: String, mode: CccdMode) = Unit

    override suspend fun nextNotification(characteristic: String): ByteArray = byteArrayOf()

    override suspend fun sleep(ms: UInt) = Unit
}

class PtpExecutorTransportStub : PtpExecutorTransport {
    override suspend fun reserveTransactionId(): UInt = 1u

    override suspend fun sendCommandFrame(frame: ByteArray) = Unit

    override suspend fun nextCommandFrame(): ByteArray = byteArrayOf()

    override suspend fun nextEventFrame(eventCode: UShort): ByteArray = byteArrayOf()

    override suspend fun openChannel(role: SocketRole) = Unit

    override suspend fun closeCommandChannel(transportCloseFrame: ByteArray?) = Unit

    override suspend fun reopenCommandSession(): PtpSessionOpenResult =
        PtpSessionOpenResult(1u, 0x2001u.toUShort(), emptyList())

    override suspend fun sleep(ms: UInt) = Unit
}

class StepObserverStub : StepObserver {
    override fun onStep(report: StepReport) = Unit
}

class ConnectionActivityObserverStub : ConnectionActivityObserver {
    override fun onActivity(event: ConnectionActivityEvent) = Unit
}

fun stepFailureDetail(error: ExecutorException.StepFailed): String =
    "${error.kind}:${error.detail}:${error.context.size}"

fun classifyFailure(kind: ExecutorStepFailureKind): Boolean = when (kind) {
    ExecutorStepFailureKind.DEADLINE_EXCEEDED -> false
    ExecutorStepFailureKind.CONDITION_REJECTED -> true
    ExecutorStepFailureKind.OTHER -> false
}

fun activityRetryCount(event: ConnectionActivityEvent): UInt = when (event) {
    is ConnectionActivityEvent.Started -> 0u
    is ConnectionActivityEvent.Retrying -> event.retry.ordinal
    is ConnectionActivityEvent.Succeeded -> event.summary.retryCount
    is ConnectionActivityEvent.Failed ->
        event.summary.retryCount + event.failure.context.size.toUInt()
    is ConnectionActivityEvent.Cancelled -> event.summary.retryCount
}

fun activityRetryBindingShape(): ConnectionActivityRetry = ConnectionActivityRetry(
    ordinal = 2u,
    limit = 3u,
    failure = ConnectionActivityFailure(
        kind = ExecutorStepFailureKind.CONDITION_REJECTED,
        context = emptyList(),
    ),
)

fun retryBindingShape(): Step = Step.Retry(
    steps = emptyList(),
    whenFailure = ExecutorStepFailureKind.CONDITION_REJECTED,
    onFailure = emptyList(),
    retryWhen = Predicate("detail", PredicateOp.EQ, "2"),
    maxAttempts = 2u,
    retryDelayMs = 200u,
    failureContext = listOf("detail"),
)

File: ci/swift/ExecutorBindingsStub.swift
import Foundation

final class ConnectionActivityObserverStub: ConnectionActivityObserver, @unchecked Sendable {
    func onActivity(event: ConnectionActivityEvent) {}
}

final class PtpExecutorTransportStub: PtpExecutorTransport, @unchecked Sendable {
    func reserveTransactionId() async throws -> UInt32 { 1 }
    func sendCommandFrame(frame: Data) async throws {}
    func nextCommandFrame() async throws -> Data { Data() }
    func nextEventFrame(eventCode: UInt16) async throws -> Data { Data() }
    func openChannel(role: SocketRole) async throws {}
    func closeCommandChannel(transportCloseFrame: Data?) async throws {}
    func reopenCommandSession() async throws -> PtpSessionOpenResult {
        PtpSessionOpenResult(transactionId: 1, responseCode: 0x2001, responseParams: [])
    }
    func sleep(ms: UInt32) async throws {}
}

func classifyFailure(_ kind: ExecutorStepFailureKind) -> Bool {
    switch kind {
    case .deadlineExceeded, .other:
        return false
    case .conditionRejected:
        return true
    }
}

func stepFailureContext(_ error: ExecutorError) -> [KeyValue] {
    guard case .StepFailed(step: _, kind: _, detail: _, context: let context) = error else {
        return []
    }
    return context
}

func activityRetryCount(_ event: ConnectionActivityEvent) -> UInt32 {
    switch event {
    case .started:
        return 0
    case let .retrying(_, _, retry):
        return retry.ordinal
    case let .succeeded(_, _, summary), let .cancelled(_, _, summary):
        return summary.retryCount
    case let .failed(_, _, summary, failure):
        return summary.retryCount + UInt32(failure.context.count)
    }
}

func activityRetryBindingShape() -> ConnectionActivityRetry {
    ConnectionActivityRetry(
        ordinal: 2,
        limit: 3,
        failure: ConnectionActivityFailure(kind: .conditionRejected, context: [])
    )
}

func retryBindingShape() -> Step {
    .retry(
        steps: [],
        whenFailure: .conditionRejected,
        onFailure: [],
        retryWhen: Predicate(field: "detail", op: .eq, value: "2"),
        maxAttempts: 2,
        retryDelayMs: 200,
        failureContext: ["detail"]
    )
}
